doc/subsystems/jinjagen/jinjagen.rst

```diff
@@ -0,0 +1,223 @@
+extend or reuse a template, the template must be stored in one of the folders listed
+in the search path list. Search paths can be set using the SEARCH_PATH key word in the
+CMakeLists.txt file.
+
+Include
+-------
+
+Include works more or less like include pre-processor directive in C, and will
+insert the content in another template starting from the location of the include keyword.
+
+.. code-block:: c
+
+	{% include 'another_template.jinja2' %}
+
+
+Import
+------
+
+The import keyword will import all macros defined in another template into a namespace.
+Similar to how import in python works. 
+
+.. code-block:: c
+
+	{% import 'template_with_macros.jinja2' as macros %}
+
+	{{ macros.an_imported_macro() }}
+
+
+Extends
+-------
+
+The extends keyword is used to extend and overload macros from another template. 
+Similar to how a child class can do the same with a parent class in object orient programming.
+
+.. code-block:: c
+
+	{% extends 'base_template.jinja2' %}
+
+	{% macro added_macro() %}
+	  This macro is from the child template
+	{% endmacro %}
+
+	{% block overload_content_in_parent_template %}
+	  This is inserted into the parent template class block from the child template
+	{% endblock %}
+
+Block
+^^^^^
+
+The block keyword can be used to define a block in the parent template that can be overloaded
+by the child template. The child block can also include the content parent block by calling
+super().
+
+.. code-block:: c
+
+    {# This is the parent template #}
+
+    {% block a_block %}
+      This is the content of the parent block
+    {% endblock %}
+
+    {# This is the child template #}
+    {% extends 'parent_template.jinja2' %}
+
+    {% block  a_block() %}
+       This is the text from the parent block: {{ super() }}
+       I add some text here!
+    {% endblock %}
+
+Code Generation in the Build Process
+====================================
+
+Code generation has to be invoked as part of the build process. Zephyr uses
+`CMake <https://cmake.org/>`_ as the tool to manage building the project.
+
+A template file that must be rendered before it can be used in a project 
+by one of the following commands in a `CMakeList.txt` file.
+
+.. code-block:: c
+
+    zephyr_sources_codegen(codegen_file.c [CODEGEN_DEFINES defines..] [SEARCH_PATH target.. path..])
+
+    zephyr_sources_codegen_ifdef(ifguard codegen_file.c [CODEGEN_DEFINES defines..] [SEARCH_PATH target.. path..])
+
+    zephyr_library_sources_codegen(codegen_file.c [CODEGEN_DEFINES defines..] [SEARCH_PATH target.. path..])
+
+    zephyr_library_sources_codegen_ifdef(ifguard codegen_file.c [CODEGEN_DEFINES defines..] [SEARCH_PATH target.. path..])
+
+The arguments given by the ``CODEGEN_DEFINES`` keyword have to be of the form
+``define_name=define_value``. The arguments becomes a part of the data object
+that can be accessed by any template and can be accessed through ``data['runtime']['defines'][define_name]``
```

Comment:
Can you elaborate on when you would want to use a CODEGEN_DEFINE?

I am concerned we have too many mechanisms with overlapping functionality and would like to get rid of it if possible.

Reply:
So more or less the same reasons for using gcc -D, but for a code generation session. 

* it can be a feature toggle, 
* render multiple variants of the same template
   * Similar hardware module share the same driver, but one instance need some errata code

on the top of my head